core/deployment/src/main/java/io/quarkus/deployment/dev/ClassComparisonUtil.java

```diff
@@ -146,21 +148,38 @@ static boolean compareMethodAnnotations(Collection<AnnotationInstance> a, Collec
                 return false;
             }
         }
+        for (Map.Entry<Integer, List<AnnotationInstance>> entry : paramTypes1.entrySet()) {
+            List<AnnotationInstance> other = paramTypes2.get(entry.getKey());
+            if (!compareAnnotations(other, entry.getValue())) {
+                return false;
+            }
+        }
         return true;
     }
 
     private static void methodMap(Collection<AnnotationInstance> b, List<AnnotationInstance> method2,
-            Map<Integer, List<AnnotationInstance>> params2) {
+            Map<Integer, List<AnnotationInstance>> params2, Map<Integer, List<AnnotationInstance>> paramTypes2) {
         for (AnnotationInstance i : b) {
-            if (i.target().kind() == AnnotationTarget.Kind.METHOD) {
-                method2.add(i);
-            } else {
-                int index = i.target().asMethodParameter().position();
-                List<AnnotationInstance> instances = params2.get(index);
-                if (instances == null) {
-                    params2.put(index, instances = new ArrayList<>());
-                }
-                instances.add(i);
+            int index;
+            switch (i.target().kind()) {
+                case METHOD:
+                    method2.add(i);
+                    break;
+                case METHOD_PARAMETER:
+                    index = i.target().asMethodParameter().position();
+                    params2.computeIfAbsent(index, k -> new ArrayList<>()).add(i);
+                    break;
+                case TYPE:
+                    TypeTarget.Usage usage = i.target().asType().usage();
+                    if (usage == TypeTarget.Usage.METHOD_PARAMETER) {
```

Comment:
I'm fairly sure at least `Usage.EMPTY` is quite realistic (type annotation on return type), and there's a few other kinds of usages that may legitimately occur. But I agree we can deal with those as they come.